### ISSUE TITLE
Support novideo while tranferring to another room

### ISSUE THREAD

**Issue by tim-cools** [steveseguin/obsninja#672]
Hi again! 

I'm trying to modify the code for the following scenario:

**transfer with novideo**
1. create two Group Chats a & b
2. a user joins group chat a
3. the director transfers the user from a to b with novideo enabled (I'll add a new button for that)
4. the user is in group b, but with novideo enabled (with optional stream id's)
5. When transferred back with the original button the novideo is disabled again

I looked into the modification and I see that  a *migrate* message is sent and a *transferred* message is received in the user's browser. Few questions:
1. I assume you need to update the wss endpoint to propagate a new field *noVideo* from the *migrate* message *transferred* message?
2. Is it possible to set **session.novideo** once the session is connected?
3. Is there any way to hook into the code when the *transferred*, so that I can set session.novideo, is executed without deobfuscating the webrtc.js file?

**hack**

A hack I can try is to abuse the chat message to notify the user that session.novideo should be set or reset. Do you think it will work to update session.novideo once the session is connected and the stream is running? 
I tried this briefly, and it doesn't seem to work out of the box...

```
function enableNoVideo(streamIds) {
	if (!streamIds){
		session.novideo = [];
	} else {
		session.novideo = streamIds.split(",");
	}
	session.connect();
}

function disableNoVideo() {
	session.novideo=null;
	session.connect();
}

```

If not would it be possible to refresh the page automatically with the *novideo*¨ parameter? I tried it briefly, and it seems that the used is going to the original room again and has to select the webcam again.

** session send function **

Can you also please explain the difference between session.sendPeers, session.sendMsg, session.sendMessage and session.sendRequest?

I would be open to contribute to a more sustainable solution, but the conference I'm developing this for is next week, hence the hacks I'm proposing ;-)

Many thanks for any feedback?

**Comment by steveseguin** [comment]
If you refresh the page, the client will revert back to the room specified in the URL.  The same goes for the case of them losing their Internet connection.  To get around this though, you can update the room URL before reloading.  You can do that by running:

```
updateURL("room=ROOM123", force=true);
``

You can also put the guest into an IFRAME and use a custom API to update the IFRAME with a new room ID and refresh it. This is how I recommend most users develop around OBS.Ninja.

&webcam&autostart , if added to the URL, will have the guest auto-join with camera already selected.   This gives you the option to also add &novideo and any other parameter, and it's more reliable the the current transfer method.

The transfer method will be revisited in the coming months, as it was designed for security, not reliability. The guest does not know what room they are transfer to, (done on purpose), and so they must be re-added if disconnected or refreshed.

You can update session.novideo dynamically; it will start working from that point forward, but not disconnect videos that are already connected.  IF you set it BEFORE you transfer a guest, it should work fine, as the transfer function disconnects existing streams.

For reference, this is the logic that occurs every time a user is requesting a new stream.  
```
if (session.novideo !== false){
   if (session.novideo.includes(session.rpcs[UUID].streamID)){
      msg.video = true;
   } else {
      msg.video = false;
}
```
==========

As per messaging, you can try to exploit the chat, but my goal is to encourage people to use the IFRAME API to handle messaging and controls. 

sendPeers = will send a json msg to connected peers; both viewers and publishers. If someone is both a viewer and pusher, it will send just once to that peer, instead of twice.

sendRequest = will send a json msg to connected publishers

sendMessage = will send a json msg to connect viewers.

Setting a UUID for the above will target a single connection, versus all that match.

As for sendMsg, this is websocket-based messaging and this shouldn't be used. I'm doing everything I can to move away from any server-side code, but this is a necessary evil for now.

**Comment by steveseguin** [comment]
I can try to add this feature in to the main code; I've recently added something similar for a call-in show format.  Might need to understand the context of the use case and problem better though.

**Comment by tim-cools** [comment]
Sure, the use case: it's an online conference with 5 stages, each stage has 5 ninja rooms: 
- waiting: people are always invited to this room first, then transferred to one of the following rooms
- production: people can only see the master feed and send audio, but do not broadcast their video to other people in the room. Only the view stream is used.
- solo1 & solo2: 2 rooms to test people's connectivity and troubleshoot for technical problems
- guest: a room where people can talk to each other

How I implemented it, the conference is currently running: I implemented custom messages which use the chat system used to move people to a different room. Depending on the room different parameters are used in the redirect. This is the current logic of the move:

```
  function moveRoom(room, broadcast) {

	var params = parseQueryString(location.search);
	params['room'] = room;

	if (session.title === "ScreenShare") {
		params['screenshare'] = true;
	} else {
		params['webcam'] = true;
	}

	params['autostart'] = true;

	if (broadcast && broadcast !== 'false') {
		params['broadcast'] = null;
	} else {
		delete params['broadcast'];
	}

	if (session.password) {
		params['password'] = session.password;
		delete params['hash'];
	}

	window.location.href = `/${buildQueryString(params)}`;
  }
```

I hard coded the rooms and the parameters used in dropdown boxes, so that they can't send someone to a wrong room. A system to define the available rooms would be nice, but I know it its very customized...

It's not the most elegant solution, but it's working for now ;-)
